Adds python code for generating base layer

evasinha-pnnl opened the pull request:
Min Chen’s Matlab script converted to python

Fixes #5

evasinha-pnnl committed:
Adds python code for generating base layer

Min Chen’s Matlab script converted to python

Fixes #5

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+    :param data:             subset(only land cells) of coarse grid coordinates along with values for variable
+    :param coarse_grid_coord pandas dataframe containing latitude and longitude coordinate of coarse grid
+    :param fine_grid_coord   pandas dataframe containing latitude and longitude coordinate of fine grid
+    
```
Need to document what the function is returning by inserting `:return:     <your explanation>` after your parameter descriptions.

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+
+    # Interpolate from coarse grid to finer resolution grid. 
+    # Convert dataframe to ndarray for method='nearest' to work
+    data_fine = griddata(points=coarse_grid_data[['Latcoord','Loncoord']].values, 
```
'Latcoord' and 'Loncoord' need to be variables passed into the function - something like `latitude` and `longitude`.  Our new base layers in Demeter now use 'latitude' and 'longitude' as the field names.  You can set the values you are passing as defaults if you would like by using:

```python
def interpolate_coarse_fine(data, coarse_grid_coord, fine_grid_cooord, latitude="Latcoord", longitude="Loncoord"):
```

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+                         method='nearest')
+    
+    return data_fine
+# ------------------------------------------------------------------------
```
Remove "# ------"

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+    return data_fine
+# ------------------------------------------------------------------------
+
+project_dir = '/pic/projects/im3/lulc_thrust_area/minchen/Demeter_0.05deg_trial/Demeter_0.05deg_trial'
```
So here is where packaging Python becomes a bit different than writing for scripting purposes only.  We want classes or functions in our code that will be called by other scripts.  So this section on should be wrapped in a class with methods, or a function or series of functions. All of your hard-coded paths will need to be removed.  You can add these and the call to your code into my tutorial notebook here: https://github.com/IMMM-SFA/clm_demeter/blob/master/tutorial.ipynb

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+# Base layer file for Plant Function Type (PFT)
+# baselayer_pft_file = os.path.join(project_dir, 'mksrf_78pft_landuse_rc2000_c130927.nc')
+baselayer_pft_file = 'mksrf_78pft_landuse_rc2000_c130927.nc'
+
```
I think you could make all of this script a class with methods.  I can give you examples of this if you would like.

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+dataset = Dataset(baselayer_pft_file, mode='r') # file handle, open in read only mode
+
+# Read latitude, longitude values
+lat = dataset.variables['LAT']  # dim = lat
```
All hardcoded values, unless they are strictly used for indexing or slicing (still need to be documented in the comments which you have done in most cases), need to be passed as variables into a function or class containing this code.  Example in this case is the use of 'LAT'.  I will not document the rest to save time.

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+# Read data for PFT variable [pft (79), lat (3600), lon (7200)]
+pft = dataset.variables['PCT_PFT']
+
+close the file
```
Missing comment #

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+pft      = pft[:,::-1,:] 
+
+# PFT values are converted from percentage (0-100) to fraction (0-1)
+pft = pft*0.05*0.05*0.01; # unit conversion
```
For Python, include spacing between operations (e.g., `*`).  Explain these multipliers and pass them as variables.

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+fid = np.array(range(np.shape(lat)[0] * np.shape(lon)[0]))
+
+# Generate PFT column names
+PFT_names = ['PFT'+ str(x) for x in range(1,pft.shape[1]+1)]
```
For Python, need space between commas for arguments (e.g. `range(1, pft.shape[1] + 1)]`

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+pft = pd.DataFrame(pft, columns=PFT_names)
+
+# Round lat and lon values to 3 decimal digits
+pft.Latcoord = pft.Latcoord.round(3)
```
Better use this type of field referencing so you can utilize variables: 

```python
pft[latitude] = pft[latitude].round(3)
```
, where `latitude` is your variable that would have the value of 'Latcoord'

Why was `3` the chosen number to round to? @chenminbnu

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+fine_grid_coord = pd.DataFrame({'Latcoord':lat_xy, 'Loncoord':lon_xy}).round(3)
+
+# Coarse grid coordinates
+[lat_coarse, lon_coarse] = np.meshgrid(np.arange(-89.875,89.875,0.25), np.arange(-179.875,179.875,0.25)) # 0.25 degree meshgrid
```
Need variables for these coords

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,161 @@
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt # to generate plots
+from netCDF4 import Dataset     # to work with NetCDF files
```
These need to be added into our `requirements.txt` file if they are not present so they are installed when the user installs the Python package we are building.

crvernon replied:
Also, `netCDF4` requires a C dependency that may complicate things a bit but we will test this.

evasinha-pnnl replied:
I don't need to import matplotlib.pyplot. I used it to temporarily check the netcdf variables read and whether flipud worked correctly. I will remove the line in the next round of edits.

crvernon reviewed (changes requested):
@evasinha Great job for your first Python conversion!!  I am going to conduct an initial review just for style and scope related items and request changes.  Once those are made, I will re-review for functionality.  My comments will appear inline.  Thanks!

evasinha-pnnl committed:
Updated script based on feed back by Chris Vernon

Converted code to class with methods
Hard coded file paths were removed
Added variables for coordinates

evasinha-pnnl commented:
@crvernon I made a bunch of edits to address most of the issues. Please review the updated code.

evasinha-pnnl committed:
Added requirements for netCDF4

evasinha-pnnl committed:
Added input files required for prepare_baselayer

evasinha-pnnl committed:
Script to develop transition priorities among FLTs

Fixes #6
Based on note provided by Min Chen

crvernon commented on clm_demeter/flt_transition_priority.py:
```diff
@@ -0,0 +1,194 @@
+            
+        return (lccs)
+    
+    def count_transition_cells(self, lccs_t1, lccs_t2, val_1, val_2):
```
Since you are not using a class attribute that has been previously initialized you can make this method static:
```python
@staticmethod
def count_transition_cells(lccs_t1, lccs_t2, val_1, val_2):
```
Note the use of the `@staticmethod` tag and dropping the self argument being passed.  The static method knows nothing about the class and just deals with the parameters.  Which is fine for us here.

crvernon commented on clm_demeter/flt_transition_priority.py:
```diff
@@ -0,0 +1,194 @@
+import numpy as np
+import numpy.ma as ma           # masked array module
+from netCDF4 import Dataset     # to work with NetCDF files
+import os                       # operating system interface
```
Standard Python modules should be listed at the top of the import stack.  Generally the order should be as follows:
1. standard library modules
2. third-party library modules (anything installed in Python’s site-packages directory) 
3. locally-developed modules

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,293 @@
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt # to generate plots
```
This still needs to be removed

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,293 @@
+import numpy as np
+import matplotlib.pyplot as plt # to generate plots
+from netCDF4 import Dataset     # to work with NetCDF files
+import os                       # operating system interface
```
You have a commented line in the `read_baselayer` method that uses `os`, otherwise it is not used.  If the commented line need be removed, remove the import of `os` as well.

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,293 @@
+        :return                  return numpy array for pft, latitude, longitude, and FID 
+        """
+        
+        #project_dir = '/pic/projects/im3/lulc_thrust_area/minchen/Demeter_0.05deg_trial/Demeter_0.05deg_trial'
```
Remove these commented lines if they are not being used

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,293 @@
+
+        # Generating land see mask
+        # pft_tot = np.sum(pft, axis=0) # Sum values for all PFTs
+        # landmask = (pft_tot > 0)      # identify land grid cells
```
Remove comment if not used

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,293 @@
+
+        # The lines below for changing type to integer do not work since there are NaN values in few rows
+        # Change data type for FID, regAEZ, region_id, and aez_id
+        #pft_lu_nv_sn.FID = pft_lu_nv_sn.FID.astype('int64')
```
Remove this commented block if it is not being used

crvernon commented on clm_demeter/prepare_baselayer.py:
```diff
@@ -0,0 +1,293 @@
+__email__ = 'min.chen@pnnl.gov and eva.sinha@pnnl.gov'
+__copyright__ = 'Copyright (c) 2017, Battelle Memorial Institute'
+
+def interpolate_coarse_fine(data, coarse_grid_coord, fine_grid_coord, variable, latitude="latitude", longitude="longitude"):
```
Is there a reason this was not made a method in the `PrepareBaseLayer` class?

crvernon reviewed (changes requested):
Excellent work so far!  I requested changes inline.  Please also provide me a path to test data that I can use to test your work.

Thanks!

evasinha-pnnl committed:
Updated method to staticmethod

and removed python modules not used

evasinha-pnnl committed:
Moved interpolate_coarse_fine method

To be part of PrepareBaseLayer class
Removed unused python modules
Removed commented lines

evasinha-pnnl committed:
Develops transition priorities among CLM types

Fixes #6
Based on notes provided by Min Chen

evasinha-pnnl committed:
Minor edit to filename and FLT names

evasinha-pnnl committed:
Added input files required for:

flt_transition_priority.py
clm_transition_priorities.py

evasinha commented:
> Excellent work so far! I requested changes inline. Please also provide me a path to test data that I can use to test your work.
> 
> Thanks!

evasinha-pnnl commented:
Chris,

I closed the pull request by mistake and am reopening it.

I have updated the `tutorial.ipynb` to add file paths and names of input files. I am also pasting the paths below:

project_dir_CLM = `/pic/projects/im3/lulc_thrust_area/minchen/Demeter_0.05deg_trial/Demeter_0.05deg_trial`

#### original CLM netcdf file
orig_baselayer_netcdf_file = os.path.join(project_dir_CLM, `mksrf_78pft_landuse_rc2000_c130927.nc`)

#### ESACCI land cover file
esacci_lu_file = os.path.join(project_dir_CLM, `ESACCI-LC-aggregated-0.25Deg-merged1992-base.csv`)

#### Nutrient availability file
nut_avail_file = os.path.join(project_dir_CLM, `000_nutrientavail_hswd_NA_0.25.csv`)

#### Soil quality file
soil_qual_file = os.path.join(project_dir_CLM, `001_soilquality_hswd_0.25.csv`)

project_dir_ESACCI = `/pic/projects/im3/lulc_thrust_area/minchen/ESA_CCI_LandCoverData`

#### File name storing ESACCI land classes
esacci_to_flt_file = os.path.join(project_dir_ESACCI, `estimate_transitionpriority`, `lccs2pft.csv`)

#### ESACCI land use filename start
fname_sta = `ESACCI-LC-L4-LCCS-Map-300m-P1Y-`

#### ESACCI land use filename end
fname_end = `-v2.0.7.nc`

#### Start and end year for ESACCI files
sta_yr = 1992
end_yr = 1994

#### Land use variable name in ESACCI netCDF files
lu_var = `lccs_class`

project_dir_transition_prior = `/pic/projects/GCAM/evasinha/clm_demeter`

#### File storing transition priorities for geographical factors
geo_trans_prior_file   = os.path.join(project_dir_transition_prior, `transition_priority_Geo.csv`)

#### File storing transition priorities for FLTs. These will be created by FLTTransitionPriorities class
FLT_trans_prior_file   = os.path.join(project_dir_transition_prior, `transition_priority_FLTs.csv`)

#### File storing transition priorities for leaf shape factor
leaf_trans_prior_file  = os.path.join(project_dir_transition_prior, `transition_priority_Leaf.csv`)

#### File storing transition priorities for  phenology factor
pheno_trans_prior_file = os.path.join(project_dir_transition_prior, `transition_priority_Pheno.csv`)

#### File storing transition priorities for crop type factor
crop_trans_prior_file  = os.path.join(project_dir_transition_prior, `transition_priority_CropType.csv`)

#### File storing transition priorities for water management factor
water_trans_prior_file = os.path.join(project_dir_transition_prior, `transition_priority_Water.csv`)

#### File storing regrouped CLM types
clm_type_factors_file  = os.path.join(project_dir_transition_prior, `CLM_type_factors.csv`)

chenminbnu commented on clm_demeter/clm_transition_priorities.py:
```diff
@@ -0,0 +1,120 @@
+__copyright__ = 'Copyright (c) 2017, Battelle Memorial Institute'
+
+class CLMTransitionPriorities:
+    """Develop transition priorities for CLM types
```
It would be better to describe what each factor represents for in the comments.
e.g., geographical factor: Boreal, Temperate, tropical etc
leaf shape: broadleaf, needleleaf etc

evasinha-pnnl replied:
Thats a good point @chenminbnu. I will add more comments describing the factors once I receive comments from @crvernon also.

chenminbnu commented on clm_demeter/clm_transition_priorities.py:
```diff
@@ -0,0 +1,120 @@
+            trans_prior_tmp['Factor_CropType'] = self.transition_priority_lookup(self.crop_trans_prior,  'Factor_CropType', i)
+            trans_prior_tmp['Factor_Water']    = self.transition_priority_lookup(self.water_trans_prior, 'Factor_Water', i)
+        
+            # Sort DataFrame by multiple columns
```
One can specify the priority order of the factors for sorting
